Sources/TuistPlugin/Service/PluginService.swift

```diff
@@ -0,0 +1,101 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+/// A protocol defining a service for interacting with plugins.
+public protocol PluginServicing {
+    /// Loads the plugins and returns them.
+    /// - Throws: An error if couldn't load a plugin.
+    /// - Returns: The loaded `Plugins` representation.
+    func loadPlugins() throws -> Plugins
+}
+
+/// A default implementation of `PluginServicing`.
+public class PluginService: PluginServicing {
+    private let config: Config
+    private let graphLoader: GraphLoading
+    private let fileHandler: FileHandling
+    private let gitHandler: GitHandling
+
+    /// Creates a `PluginService`.
+    /// - Parameters:
+    ///   - config: The loaded `Config` used to parse plugins.
+    ///   - graphLoader: A graph loader for loading plugins.
+    ///   - fileHandler: A file handler for creating plugin directories/related files.
+    ///   - gitHandler: A git handler for cloning and interacting with remote plugins.
+    public init(
+        config: Config,
+        graphLoader: GraphLoading,
+        fileHandler: FileHandling = FileHandler.shared,
+        gitHandler: GitHandling = GitHandler()
+    ) {
+        self.config = config
+        self.graphLoader = graphLoader
+        self.fileHandler = fileHandler
+        self.gitHandler = gitHandler
+    }
+
+    public func loadPlugins() throws -> Plugins {
+        let pluginPaths = try fetchPlugins()
+        let pluginManifests = try pluginPaths.map(graphLoader.loadPlugin)
+
+        var projectDescriptionHelpers = [CustomProjectDescriptionHelpers]()
+
+        zip(pluginManifests, pluginPaths).forEach { plugin, path in
+            logger.debug("Loading plugin: \(plugin.description) at: \(path)")
+
+            switch plugin {
+            case let .helpers(name):
+                projectDescriptionHelpers.append(.init(name: name, path: path))
+            }
+        }
+
+        return .init(projectDescriptionHelpers: projectDescriptionHelpers)
+    }
+
+    private func fetchPlugins() throws -> [AbsolutePath] {
+        try config.plugins
+            .map { plugin in
+                switch plugin {
+                case let .local(path):
+                    return AbsolutePath(path)
+                case let .gitWithBranch(url, id),
+                     let .gitWithTag(url, id),
+                     let .gitWithSha(url, id):
+                    return try fetchGitPlugin(at: url, with: id, fingerprint: plugin.fingerprint)
+                }
+            }
+    }
+
+    /// fetches the git plugins from the remote server and caches them in
+    /// the Tuist cache with a unique fingerprint
+    private func fetchGitPlugin(at url: String, with gitId: String, fingerprint: String) throws -> AbsolutePath {
+        let pluginDirectory = Environment.shared.cacheDirectory
+            .appending(.init(Constants.PluginDirectory.name))
```

Comment:
I would not use `.init` as it's a little bit hard to read and derive the type just by reading this (I generally try to keep usage of `.init` low as it makes the code a little bit less explicit)